patient-dash-backend/src/index.ts

```diff
@@ -15,28 +18,119 @@ import { authApiDetail, patientApiDetail, providerApiDetail, swaggerDocumentatio
+    }, authApiDetail)
   })
   .group("/api", app => app
-    .get("/patients", () => "GET Patients", patientApiDetail)
-    .post("/patients", () => "POSTing a new patient record", patientApiDetail)
-    .get("/patients/:id", ({params: id}) => `GET Patient by ID ${id}`, patientApiDetail)
-    .patch("/patients/:id", ({params: id}) => `PATCHing patient ${id}`, patientApiDetail)
+    .get("/patients", async ({query: {
+      count,
+      page,
+      searchString
+    }}) => {
+      // Prevent unnecessary query if page requested goes beyond total page count
+      const finalPage = Math.ceil(await prisma.patient.count() / count);
+      if(page && finalPage > page) {
+        return {
+          currentPage: finalPage,
+          finalPage,
+          results: []
+        };
+      }
+
+      const res = await prisma.patient.findMany({
+        take: count,
+        skip: count * ((page || 1) - 1),
+        ...(searchString ? {where: {
+          OR: [
+            {
+              firstName: {
+                contains: searchString,
+                mode: 'insensitive'
+              }
+            },
+            {
+              lastName: {
+                contains: searchString,
+                mode: 'insensitive'
+              }
+            },
+            {
+              middleName: {
+                contains: searchString,
+                mode: 'insensitive'
+              }
+            }
+          ]
+        }} : {})
+      });
+
+      return {
+        currentPage: page || 1,
+        finalPage,
+        results: res,
+      }
+    }, {
+      ...patientApiDetail,
+      query: t.Object({
+        count: t.Number(),
+        page: t.Optional(t.Number()),
+        searchString: t.Optional(t.String()),
+        filters: t.Optional(t.Array(
+          t.Object({
+            filterKey: t.String(),
+            filterValue: t.String(),
+          })
+        ))
+      })
+    })
+    .post("/patients", ({body: {
+      firstName,
+      lastName,
+      dateOfBirth,
+      middleName,
+    }}) => {
+      const res = prisma.patient.create({
+        data: {
+          firstName,
+          lastName,
+          dateOfBirth,
+          ...(middleName ? { middleName } : {})
+        }
+      });
+      return res;
+    }, {
+      ...patientApiDetail,
+      body: t.Object({
+        firstName: t.String(),
+        lastName: t.String(),
+        dateOfBirth: t.Date(),
+        middleName: t.Optional(t.String())
+      })
+    })
+    .get("/patients/:id", ({params: { id }}) => {
+      return prisma.patient.findUnique({
+        where: {
+          id: id
+        }
+      })
+    }, {
+      ...patientApiDetail,
+      params: t.Object({
+        id: t.Number(),
+      })
+    })
+    .patch("/patients/:id", ({params: id}) => {
+      // TODO - Support adding arbitrary data, editing
+    }, patientApiDetail)
     .get("/providers/:id", ({params: id}) => `GET provider by ID ${id}`, providerApiDetail)
   )
   .listen(3001);
```

Comment:
Method chaining for type safety is pretty nifty, but I can see this getting out of hand verrrrry quickly. I'd like to break this apart before it gets too messy, but haven't found a clear path forward just yet. Thinking I can make some folder structure that matches the API layer and fill each file with most of the code needed for handling a request.